analysis/priestholy/src/modules/shadowlands/items/FlashConcentration.tsx

```diff
@@ -0,0 +1,228 @@
+      Events.applybuffstack.by(SELECTED_PLAYER).spell(SPELLS.FLASH_CONCENTRATION),
+      this.onByPlayerFlashConcentrationBuff,
+    );
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.FLASH_CONCENTRATION),
+      this.onByPlayerFlashConcentrationInitialBuff,
+    );
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.FLASH_CONCENTRATION),
+      this.onByPlayerFlashConcentrationBuffRemoval,
+    );
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.FLASH_HEAL),
+      this.onByPlayerFlashHealCast,
+    );
+    this.addEventListener(
+      Events.heal.by(SELECTED_PLAYER).spell(SPELLS.GREATER_HEAL),
+      this.onByPlayerHealAmount,
+    );
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.GREATER_HEAL),
+      this.onByPlayerHealCast,
+    );
+  }
+
+  /**
+   * Sets _isFullStack to true if there are now 5 stacks, as well as the last refresh
+   * NB: it comes after the Flash Heal Cast
+   * @param event indicate the new number of stacks (only positive increments)
+   */
+  onByPlayerFlashConcentrationBuff(event: ApplyBuffStackEvent) {
+    if (event.stack === 5) {
+      this._isFullStack = true;
+      this._lastFullStackRefresh = event.timestamp;
+    } else {
+      this._isFullStack = false;
+    }
+    this._currentStacks = event.stack;
+  }
+
+  /**
+   * Sets _isFullStack to false.
+   * Note: The only exception is the prepull event, for which we assume that
+   * it means the character has five stacks. It will be corrected if the
+   * first Flash Heal generates a ApplyBuffStack or ApplyBuff event
+   * Note 2: Technically, the only time this function does something is on the first
+   * apply buff if there was no prepull FC.
+   * @param event with info about first stack, or prepull incomplete indication
+   */
+  onByPlayerFlashConcentrationInitialBuff(event: ApplyBuffEvent) {
+    if (event.prepull !== true) {
+      this._isFullStack = false;
+      this._currentStacks = 1;
+    }
+  }
+
+  /**
+   * Increments _totalDrops for our suggestion, since that should ideally not happen
+   */
+  onByPlayerFlashConcentrationBuffRemoval(event: RemoveBuffEvent) {
+    this._totalDrops += 1;
+    this._isFullStack = false;
+    this._currentStacks = 0;
+  }
+
+  /**
+   * when at full stack, it refreshes the last refresh timer,
+   * and counts a Wasteful Cast if it was used earlier than needed
+   */
+  onByPlayerFlashHealCast(event: CastEvent) {
+    if (this._isFullStack) {
+      if (event.timestamp - this._lastFullStackRefresh < 15000) {
+        this._wastefulCasts += 1;
+      }
+      this._lastFullStackRefresh = event.timestamp;
+    }
+  }
+
+  onByPlayerHealAmount(event: HealEvent) {
+    this._totalHeal += event.amount;
+    this._gainedHeal +=
+      (event.amount * (this._currentStacks * 0.03)) / (1 + this._currentStacks * 0.03);
+    this._gainedCastTime += this._currentStacks * 200;
```

Comment:
All spell or analysis related constant numbers should be at the top of the file above the class instead of hardcoded. This makes it easier to see what the numbers refer to and also easier to update if a spell gets buffed or nerfed. In this case the `0.03` and `200` should be moved into consts.

See Flourish.tsx for an example of this.